mlos_bench/mlos_bench/services/remote/ssh/ssh_service.py

```diff
@@ -0,0 +1,379 @@
+    # Using a larger thread pool requires a bit more work since asyncssh
+    # requires that run() requests are submitted to the same event loop handler
+    # that the connection was made on.
+    # In that case, each background thread should get its own SshClientCache.
+
+    _event_loop: Optional[AbstractEventLoop] = None
+    _event_loop_thread: Optional[Thread] = None
+    _event_loop_thread_lock = ThreadLock()
+    _event_loop_thread_refcnt: int = 0
+    _event_loop_thread_ssh_client_cache: Optional[SshClientCache] = None
+
+    _REQUEST_TIMEOUT: Optional[float] = None  # seconds
+
+    def __init__(self,
+                 config: Optional[Dict[str, Any]] = None,
+                 global_config: Optional[Dict[str, Any]] = None,
+                 parent: Optional[Service] = None,
+                 methods: Union[Dict[str, Callable], List[Callable], None] = None):
+        # pylint: disable=too-complex
+
+        super().__init__(config, global_config, parent, methods)
+
+        # Make sure that the value we allow overriding on a per-connection
+        # basis are present in the config so merge_parameters can do its thing.
+        self.config.setdefault('ssh_port', None)
+        assert isinstance(self.config['ssh_port'], (int, type(None)))
+        self.config.setdefault('ssh_username', None)
+        assert isinstance(self.config['ssh_username'], (str, type(None)))
+        self.config.setdefault('ssh_priv_key_path', None)
+        assert isinstance(self.config['ssh_priv_key_path'], (str, type(None)))
+
+        # None can be used to disable the request timeout.
+        self._request_timeout = self.config.get("ssh_request_timeout", self._REQUEST_TIMEOUT)
+        self._request_timeout = float(self._request_timeout) if self._request_timeout is not None else None
+
+        # Prep an initial connect_params.
+        self._connect_params: dict = {
+            # In general scripted commands shouldn't need a pty and having one
+            # available can confuse some commands, though we may need to make
+            # this configurable in the future.
+            'request_pty': False,
+            # By default disable known_hosts checking (since most VMs expected to be dynamically created).
+            'known_hosts': None,
+        }
+
+        if 'ssh_known_hosts_file' in self.config:
+            self._connect_params['known_hosts'] = self.config.get("ssh_known_hosts_file", None)
+            if isinstance(self._connect_params['known_hosts'], str):
+                known_hosts_file = os.path.expanduser(self._connect_params['known_hosts'])
+                if not os.path.exists(known_hosts_file):
+                    raise ValueError(f"ssh_known_hosts_file {known_hosts_file} does not exist")
+                self._connect_params['known_hosts'] = known_hosts_file
+        if self._connect_params['known_hosts'] is None:
+            _LOG.info("%s known_hosts checking is disabled per config.", self.__class__.__name__)
```

Comment:
```suggestion
            _LOG.info("%s known_hosts checking is disabled per config.", self)
```
if we have `__repr__` or `__str__` defined (+ a case for human-readable `str()`)